### PR TITLE
Add more tooltips for action buttons

### LINKED ISSUE #5972: Pipeline view enhancements
## Summary

Update the current appearance/behavior of Concourse's pipeline view to achieve the following:

- [x] Create tooltips for all action buttons (play/pause, and deactivated elements like resource pin icon in pipeline view) in the top right hand part of the viewport
- [x] Ensure foreground/background colours of the action buttons mentioned above are consistent with contrast rules throughout the rest of the app for default and hover states
- [ ] Uniform padding (top, right, bottom, left) on jobs/resources in pipeline string diagrams

## Todo:

**Pipeline View state:**
Screen Shot 2020-11-23 at 1 37 00 PM

* Top right icons (pin, favorite, play/pause) get tooltips

**Current padding (12px, 14px, 12px, 14px, with 20px between icon and text):**
Screen Shot 2020-11-23 at 1 49 16 PM

**Proposed padding (8px, 8px, 8px, 8px, with 16px between icon and text):**
Screen Shot 2020-11-23 at 1 50 23 PM

## Context

* Part of #5964 Web UI/UX experience

### PR DESCRIPTION
**Pull request by aoldershaw**
[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## What does this PR accomplish?

Bug Fix | **Feature** | Documentation

closes #5972 .

Adds tooltips to the pipeline page as per the issue, as well as the Dashboard, Build page, Job page, and Resource page. Also makes tooltip colours consistent throughout the app.

Note: it does not address the uniform padding issue on the pipeline page - there were browser inconsistencies which made it difficult/impossible to make things look good across all browsers.

## Changes proposed by this PR:

* Use the new(ish) `tooltip` function for defining all tooltips (that I could find), rather than each page rendering their own tooltips
* Made it easier to render consistently styled tooltips by applying default styles in the `Tooltip` module (and consequently use the new tooltip style for all the tooltips)
* New tooltips
  * Dashboard
    * Pipeline card favourite icon
    * Pipeline card pause toggle
  * Pipeline page
    * Top bar pause toggle
    * Top bar favourite icon
    * Top bar pin menu icon
  * Build page
    * Trigger build button
    * Rerun build button
    * Abort build button
  * Job page
    * Trigger build button
    * Pause toggle button
  * Resource page
    * Trigger manual check button
    * Pin version button
    * Enable version button
    * Unpin button (in pin bar)
    * Edit comment button
    * Pin bar (for statically pinned)

## Release Note

* Many buttons in the UI now have a tooltip on hover to indicate what they do

## Contributor Checklist

- [ ] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [ ] [Signed] all commits
- [ ] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist

- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

### CODE REVIEW

**Comment by chenbh** [comment]
- if you hover over the `Re-run Build` or `Trigger Build` buttons long enough, the default HTML/CSS tooltip appears 
- can you also add tooltips for favouriting pipelines in the sidebar?
- the expose button on the pipeline card updates inconsistently after clicking, I would expect the icon and tooltip to change immediately after I click it

https://user-images.githubusercontent.com/17577073/106659780-2fa33d80-656d-11eb-8d2a-32be48165ad4.mov

- what do you think of adding tooltips for going jobs <-> builds page? It's not well known that you can click on the name in the builds page to go to jobs page and vice versa. (We also never directly refer to them as jobs/builds page anywhere in the UI)
- is it worth it to add tooltips to the left/right chevrons to indicate prev/next page?

**Comment by aoldershaw** [comment]
@chenbh all good feedback, thanks! addressed all your comments, except for

> can you also add tooltips for favouriting pipelines in the sidebar?

It may be possible, but tbh with the way tooltips are handled now, it's not easy to do - our current hover behaviour can't really deal with overlapping elements that can both be hovered

So, if it's cool with you, I think I'll just leave the tooltip off on the sidebar favourited icon for now

**Review by chenbh — COMMENTED** [review]
I think we'll need to set `hovered = HoverState.NoHover` on generic updates to the model as well. When I favourite a pipeline from the dashboard view, the viewport jumps down but the tooltip is still left behind.

https://user-images.githubusercontent.com/17577073/107250091-cce4f280-6a01-11eb-92f3-5602502db6e1.mov


 https://github.com/concourse/concourse/blob/370b34f7524696627c0047da39a71c2d114f9dd6/web/elm/src/Application/Application.elm#L297-L315

**Comment by chenbh** [comment]
Not part of this PR, but we are missing a bunch of accessibility labels:

Resource page:
- re-check button
- pin bar unpin button (same location as the re-check button)

Builds page:
- step tree (I'm not 100% sure if there's an accessible way to expand the step to view logs)
- image fetching button

Pipeline page:
- topbar show pinned resources
- topbar favourite
- topbar pause pipeline

Dashboard
- card pause pipeline
- card expose pipeline
- card favourite pipeline

cc @matthewpereira

File: web/elm/src/Build/Build.elm [hunk 1]
```diff
@@ -661,12 +661,30 @@ view session model =
...
+tryAll maybes =
+    List.foldl
+        (\prev cur ->
+            case prev of
+                Nothing ->
+                    cur
+
+                _ ->
+                    prev
+        )
```
[the remark this answers is not in the document]
Comment (aoldershaw) [reply web/elm/src/Build/Build.elm:685]: yeah - there's probably a nicer way of doing it, couldn't be bothered to figure it out at the time. I guess one way to make it clearer is `List.filter ((/=) Nothing) >> List.head`

**Comment by aoldershaw** [comment]
> I think we'll need to set `hovered = HoverState.NoHover` on generic updates to the model as well. When I favourite a pipeline from the dashboard view, the viewport jumps down but the tooltip is still left behind.

Some elements use `onMouseEnter/Leave` (which only triggers once) instead of `onMouseOver/Out`, so resetting `hovered = HoverState.NoHover` gets rid of the tooltip altogether. I'm not sure if there's a reason that they use `mouseenter` vs `mouseout`, so not sure exactly what the implications of changing them are.

Another (less general) option is to add an `onMouseOver/Out` on the `Dashboard`, which only helps for this specific case - not sure of other situations where things get shifted around

File: web/elm/src/Build/Styles.elm [hunk 2]
```diff
@@ -236,22 +236,12 @@ stepStatusIcon =
 
 changedStepTooltip : List (Html.Attribute msg)
 changedStepTooltip =
-    [ style "padding" "5px"
-    , style "z-index" "100"
-    , style "width" "fit-content"
-    , style "pointer-events" "none"
-    ]
-        ++ Application.Styles.disableInteraction
+    style "pointer-events" "none" :: Application.Styles.disableInteraction
```
[the remark this answers is not in the document]
Comment (aoldershaw) [reply web/elm/src/Build/Styles.elm:239]: It's *technically* possible to construct a tooltip that overlaps the element, though you probably would never do that. Either way, I think semantically it makes sense for the tooltip to ignore pointer events

**Comment by aoldershaw** [comment]
@chenbh I decided to take the less general but less risky approach of just adding an `onMouseOver/Out` to the `Dashboard` element to fix the "tooltip not disappearing" issue after you favourite a pipeline

**Review by chenbh — APPROVED** [review]
lgtm

**Comment by klhex** [comment]
Nice. But what about a user-configurable tooltip when hovering over a job name in the pipeline overview? I.e. a user-provided description on what a job does...

**Pull request merged**